### ISSUE TITLE
Installation of extension after uninstalling the extension, updating the code, rebuilding and installation fails

### ISSUE THREAD

**Issue by idevrlabs** [docker/extensions-sdk#210]
I made changes to the code added few functionality adding few more buttons and text boxes in App.tsx and vm main.go code 

Docker extension built and installed successfully with modified code with a simple functionality and worked fine

After successfully testing the extension with UI updated the code added few more features  unistalled the extension and succesfully rebuilt but installation failed

C:\Users\username> docker extension install companyname/extensionname:latest
Extensions can install binaries, invoke commands and access files on your machine.
Are you sure you want to continue? [y/N] y
Installing new extension "companyname/extensionname:latest"
Installing service in Desktop VM...
Setting additional compose attributes
forwarding VM socket "/run/guest-services/companyname_extensionname/extension-extensionname.sock" to "\\\\.\\pipe\\dockerDesktopPluginExtensionextensionname": open \\.\pipe\dockerDesktopPluginExtensionextensionname: Access is denied.
Removing extension companyname/extensionname:latest...
Removing extension VM service...
Extension removed from Desktop VM
VM service socket forwarding stopped
Extension UI tab extensionname removed
Warning: extension image companyname/extensionname:latest could not be removed: executing 'docker --context default image remove companyname/extensionname:latest' : exit status 1:
Error response from daemon: conflict: unable to remove repository reference "companyname/extension:latest" (must force) - container d80dbee67c7c is using its referenced image 83d705f6c706
Extension "extensionname" removed
installation could not be completed due to: forwarding VM socket "/run/guest-services/companyname_extensionname/extension-extensionname.sock"
 to "\\\\.\\pipe\\dockerDesktopPluginExtensionextensionname": open \\.\pipe\dockerDesktopPluginExtensionextensionname: Access is denied.

Steps to reproduce is 
1. uninstall the extension
  Remove the installed extension running  
   docker extension rm companyname/extensionname:latest
2. Update the code both vm (go) and reactjs App.tsx
3.  Rebuild 
    docker build -t companyname/extensionname .

First run installed perfectly. After uninstalling and rebuilding failed with the above error

When searched for the pipe the pipe doesn't even exist.

Is there a way to clean up?

**Output of `docker extension version`:**

```
Client Version: v0.2.8
Server API Version: 0.2.3
```

**Output of `docker version`:**

```
Client:
 Cloud integration: v1.0.28
 Version:           20.10.17
 API version:       1.41
 Go version:        go1.17.11
 Git commit:        100c701
 Built:             Mon Jun  6 23:09:02 2022
 OS/Arch:           windows/amd64
 Context:           default
 Experimental:      true

Server: Docker Desktop 4.11.0 (83626)
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:01:23 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.6
  GitCommit:        10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc:
  Version:          1.1.2
  GitCommit:        v1.1.2-0-ga916309
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

Windows 11

**Comment by idevrlabs** [comment]
In makefile found the update command syntax which is not mentioned in the documentation. 

update-extension: build-extension ## Update the extension
	docker extension update $(IMAGE):$(TAG)

**Comment by idevrlabs** [comment]
In makefile found the update command syntax which is not mentioned in the documentation. 

update-extension: build-extension ## Update the extension
	docker extension update $(IMAGE):$(TAG)

**Comment by idevrlabs** [comment]
docker extension update command worked once 

Second time reported 

failed to update extension status as it wasn't found under the repo name "companyname/extensionname"

Executed docker extension rm command assuming that this will cleanup and run docker extension install command even build command failed with same reported errors as above 
C:\Users\username> docker extension install companyname/extensionname:latest
Extensions can install binaries, invoke commands and access files on your machine.
Are you sure you want to continue? [y/N] y
Installing new extension "companyname/extensionname:latest"
Installing service in Desktop VM...
Setting additional compose attributes
forwarding VM socket "/run/guest-services/companyname_extensionname/extension-extensionname.sock" to "\\.\pipe\dockerDesktopPluginExtensionextensionname": open \.\pipe\dockerDesktopPluginExtensionextensionname: Access is denied.
Removing extension companyname/extensionname:latest...
Removing extension VM service...
Extension removed from Desktop VM
VM service socket forwarding stopped
Extension UI tab extensionname removed
Warning: extension image companyname/extensionname:latest could not be removed: executing 'docker --context default image remove companyname/extensionname:latest' : exit status 1:
Error response from daemon: conflict: unable to remove repository reference "companyname/extension:latest" (must force) - container d80dbee67c7c is using its referenced image 83d705f6c706
Extension "extensionname" removed
installation could not be completed due to: forwarding VM socket "/run/guest-services/companyname_extensionname/extension-extensionname.sock"
to "\\.\pipe\dockerDesktopPluginExtensionextensionname": open \.\pipe\dockerDesktopPluginExtensionextensionname: Access is denied.

**Comment by idevrlabs** [comment]
While developing use the following instructions for hot reloading... 

https://docs.docker.com/desktop/extensions-sdk/build/test-debug/#hot-reloading-whilst-developing-the-ui

**Comment by idevrlabs** [comment]
Unable to install the companyname/extensionname:latest yet. Need a cleanup command. I may have to try with a new name.

**Comment by benja-M-1** [comment]
I managed to reproduce the issue with an extension that has an invalid name (i.e foo instead of foo/bar).

It is the windows version of https://github.com/docker/extensions-sdk/issues/215

**Comment by xenoscopic** [comment]
I've managed to reproduce this issue with an [extension](https://hub.docker.com/r/mutagenio/docker-desktop-extension) (with a valid name, AFAIK) that has long-polling HTTP operations targeting the backend HTTP service.  In development so far, I've noticed that these long-polling `GET` requests will actually outlive the UI pane (so I preempt/cancel stale requests whenever the UI pane is reactivated).  Anyway, I'm wondering if perhaps these stale requests are holding open `HANDLE`s that are preventing removal of the named pipe?  I notice that I don't see this `Access is denied` issue if I never start these long-polling requests (by never activating the pane).

Go always opens Windows file `HANDLE`s in a way that prevents deletion of the file (i.e. there's no `FILE_SHARE_DELETE` flag [here](https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/syscall/syscall_windows.go;l=331)).  I'm not sure what the extension code looks like in Docker Desktop (or if the named pipe causing issues is even being used as the transport for backend HTTP requests), but that would be my best guess on the conflict here.  If a UDS is being used for this transport on macOS/Linux, that would explain why the issue doesn't manifest there (since there's nothing to block the `unlink` system call).

**Comment by benja-M-1** [comment]
Thank you @xenoscopic!  That's a good one, I'll investigate that!

**Comment by idevrlabs** [comment]
> @benja-M-1 The extension name created with idevrl/extensionname definitely it is a valid name. 

It seems it is the windows version of https://github.com/docker/extensions-sdk/issues/215. So here the workaround is to

remove the extension folder from %userprofile%/AppData/Roaming/Docker/extensions
restart Docker Destkop

These steps to clean surely helps.

I would keep this issue open as @xenoscopic issue is an interesting case to resolve but mine was also you may have to dig deep.

Now that we have a new update will check the same and revert. If it persists will report.


Thanks

**Comment by benja-M-1** [comment]
I have made some fixes, but they don't address @xenoscopic case. They should be available in 4.13 wich should land mid-october. 

We will keep this issue open until then and the new case is fix as well.

**Comment by benja-M-1** [comment]
We released a fix that should address @xenoscopic's [issue](https://github.com/docker/extensions-sdk/issues/210#issuecomment-1244440851) in 4.14.

I close this issue. If you see this again, please feel free to reopen with details :)
